grpclb_remote_balancer.go

```diff
@@ -0,0 +1,270 @@
+//  - If the new server list == old server list, do nothing and return.
+//  - Else if the new backends == old backends, generate a new grpclb picker
+//    with the new server list and the old RR picker.
+//  - Else, send updates to RR, and don't update grpclb picker. The grpclb
+//    picker will be updated when the RR updates the RR picker.
+func (lb *lbBalancer) processServerList(l *lbpb.ServerList) {
+	// TODO(bargrpclb) see the comment.
+	grpclog.Infof("lbBalancer: processing server list: %+v", l)
+	lb.mu.Lock()
+	defer lb.mu.Unlock()
+
+	lb.serverListReceived = true
+	if reflect.DeepEqual(lb.fullServerList, l.Servers) {
+		grpclog.Infof("lbBalancer: new serverlist same as the previous one, ignoring")
+		return
+	}
+	lb.fullServerList = l.Servers
+
+	var backendAddrs []resolver.Address
+	for _, s := range l.Servers {
+		if s.DropForLoadBalancing || s.DropForRateLimiting {
+			continue
+		}
+
+		md := metadata.Pairs(lbTokeyKey, s.LoadBalanceToken)
+		ip := net.IP(s.IpAddress)
+		ipStr := ip.String()
+		if ip.To4() == nil {
+			// Add square brackets to ipv6 addresses, otherwise net.Dial() and
+			// net.SplitHostPort() will return too many colons error.
+			ipStr = fmt.Sprintf("[%s]", ipStr)
+		}
+		addr := resolver.Address{
+			Addr:     fmt.Sprintf("%s:%d", ipStr, s.Port),
+			Metadata: &md,
+		}
+
+		backendAddrs = append(backendAddrs, addr)
+	}
+
+	backendsUpdated := lb.refreshSubConns(backendAddrs)
+	// If no backend was updated, no SubConn will be newed/removed. But since
+	// the full serverList was different, there might be updates in drops or
+	// pick weights(different number of duplicates). We need to update picker
+	// with the fulllist.
+	if !backendsUpdated {
+		lb.regeneratePicker()
+		lb.cc.UpdateBalancerState(lb.state, lb.picker)
+	}
+}
+
+// refreshSubConns creates/removes SubConns with backendAddrs. It returns a bool
+// indicating whether the backendAddrs are different from the cached
+// backendAddrs (whether SubConn was newed/removed).
+// Caller must hold lb.mu.
+func (lb *lbBalancer) refreshSubConns(backendAddrs []resolver.Address) bool {
+	lb.backendAddrs = nil
+	var backendsUpdated bool
+	// addrsSet is the set converted from backendAddrs, it's used to quick
+	// lookup for an address.
+	addrsSet := make(map[resolver.Address]struct{})
+	// Create new SubConns.
+	for _, addr := range backendAddrs {
+		addrWithoutMD := addr
+		addrWithoutMD.Metadata = nil
+		addrsSet[addrWithoutMD] = struct{}{}
+		lb.backendAddrs = append(lb.backendAddrs, addrWithoutMD)
+
+		if _, ok := lb.subConns[addrWithoutMD]; !ok {
+			backendsUpdated = true
+
+			// Use addrWithMD to create the SubConn.
+			sc, err := lb.cc.NewSubConn([]resolver.Address{addr}, balancer.NewSubConnOptions{})
+			if err == nil {
+				lb.subConns[addrWithoutMD] = sc // Use the addr without MD as key for the map.
+				lb.scStates[sc] = connectivity.Idle
+				sc.Connect()
+			} else {
+				grpclog.Warningf("roundrobinBalancer: failed to create new SubConn: %v", err)
```

Comment:
Eep. (How) does this happen?  Do we actually handle this correctly?

It looks like it only happens if the ClientConn is closing.  Maybe we should just eat that error in grpc and give back some kind of dummy disabled subconn to make the balancer think it's happy?  Ideally it seems like the balancer shouldn't have to deal with this situation.

Reply:
`NewSubConn()` doesn't return any usefull errors now.
It's here in case we want to add some fields to the `NewSubConnOptions`, the returned error may become more meaningful.

Reply:
Go style:

```go
if err != nil {
  grpclog.Warning...
  continue
}
lb.subConns[addrWithoutMD] = sc
...
```